torch/quantization/fx/quantize.py

```diff
@@ -407,6 +414,696 @@ def load_arg(a):
+            # except if the second dtype is float, a dequant will be inserted
+            # without an observer in convert
+            # TODO(future PR): change this so a placeholder is inserted for
+            # future dequants, to make the logic easier to understand
+            (node_dtype != torch.float) and
+            # if arg is a bool tensor or not a tensor, do not insert observer
+            (arg_dtype not in (torch.bool, None)) and
+            (is_activation and activation_is_statically_quantized(qconfig))
+        )
+
+        needs_obs = (
+            weight_needs_obs or
+            bias_needs_obs or
+            dtype_changes_and_second_dtype_not_float
+        )
+
+    else:
+        # custom flow for standalone modules
+        _sm_qconfig_dict, sm_prepare_config_dict = \
+            get_standalone_module_configs_rewrite(
+                node, modules, prepare_custom_config_dict, qconfig)
+
+        sm_input_quantized_idxs = \
+            sm_prepare_config_dict.get('input_quantized_idxs', [])
+        # for args, this is set to the index of the current arg
+        # for kwargs, this is left at None
+        cur_input_idx = None
+        for arg_idx, arg_to_check in enumerate(node.args):
+            if arg_to_check is arg:
+                cur_input_idx = arg_idx
+                break
+
+        if cur_input_idx is None:
+            needs_obs = False
+        else:
+            arg_dtype = node_name_to_target_dtype[arg.name]
+            node_dtype = torch.quint8 if cur_input_idx in sm_input_quantized_idxs \
+                else torch.float
+            needs_obs = (
+                (arg_dtype != node_dtype) and
+                (node_dtype != torch.float)
+            )
+
+    if needs_obs:
+
+        arg_this_env = load_arg(arg)
+        new_obs_mod = act_post_process_ctr()
+        existing_obs_node = None
+
+        # Before using the new observer, check if an observer
+        # of the correct type already exists. If it does, use it.
+        # This prevents duplicate observer insertions if a node is
+        # used by multiple nodes.
+        for maybe_obs_node, _ in arg_this_env.users.items():
+            if maybe_obs_node.op == 'call_module':
+                maybe_obs_mod = modules[maybe_obs_node.target]
+                if (
+                    type(maybe_obs_mod) == type(new_obs_mod) and
+                    node_name_to_target_dtype[maybe_obs_node.name] == node_dtype
+                ):
+                    existing_obs_node = maybe_obs_node
+                    break
+
+        if existing_obs_node is None:
+            new_obs_node = insert_observer_rewrite(
+                arg_this_env, new_obs_mod, model, modules, graph)
+            # set the type, so the next node can read it
+            node_name_to_target_dtype[new_obs_node.name] = node_dtype
+            # override this arg to be the observed arg
+            new_arg = new_obs_node
+        else:
+            new_arg = existing_obs_node
+
```

Comment:
if it does not need observer, looks like we would return `arg`, which is from the old graph?

Reply:
correct, this part of the code mutates the nodes of an existing graph (without creating a new graph).

if the RFC goes forward, a potential future improvement would be to stop using the `env` pattern in inserting observers for outputs as well, and mutate the graph directly.  Using env works ok for inserting observers to outputs (because we are iterating through the graph in order), but it does not work cleanly for inserting observers to node inputs (since this is out of order).

Reply:
oh I see, yeah having env and mutating the graph sounds a bit confusing